core/includes/backup/backup.file.inc

```diff
@@ -0,0 +1,260 @@
+    }
+    return NULL;
+  }
+
+  /**
+   * Close a file when we're done reading/writing.
+   */
+  public function close() {
+    fclose($this->handle);
+    $this->handle = NULL;
+  }
+
+  /**
+   * Write a line to the file.
+   */
+  public function write($data) {
+    if (!$this->handle) {
+      $this->handle = $this->open(TRUE);
+    }
+    if ($this->handle) {
+      fwrite($this->handle, $data);
+    }
+  }
+
+  /**
+   * Read a line from the file.
+   */
+  public function read($size = NULL) {
+    if (!$this->handle) {
+      $this->handle = $this->open();
+    }
+    if ($this->handle && !feof($this->handle)) {
+      return $size ? fread($this->handle, $size) : fgets($this->handle);
+    }
+    return NULL;
+  }
+
+  /**
+   * Write the entire contents of the file.
+   */
+  public function putContents($data) {
+    file_put_contents($this->filePath(), $data);
+  }
+
+  /**
+   * Read the entire contents of the file.
+   */
+  public function getContents() {
+    return file_get_contents($this->filePath());
+  }
+
+  /**
+   * Get the file size.
+   */
+  public function getFileSize() {
+    return filesize($this->filePath());
+  }
+
+  /**
+   * Transfer file using http to client. Similar to the built in file_transfer,
+   * but it calls module_invoke_all('exit') so that temp files can be deleted.
+   */
+  public function transfer() {
+    $headers = array(
+      array(
+        'key' => 'Content-Type',
+        'value' => $this->mimetype(),
+      ),
+      array(
+        'key' => 'Content-Disposition',
+        'value' => 'attachment; filename="'. $this->filename() .'"',
+      ),
+    );
+    // In some circumstances, web-servers will double compress gzipped files.
+    // This may help alleviate that issue by disabling mod-deflate.
+    if ($this->mimetype() == 'application/x-gzip') {
+      $headers[] = array(
+        'key' => 'Content-Encoding',
+        'value' => 'gzip',
+      );
+    }
+    if ($size = $this->getFileSize()) {
+      $headers[] = array(
+        'key' => 'Content-Length',
+        'value' => $size,
+      );
+    }
+
+    // Suppress the warning you get when the buffer is empty.
+    @ob_end_clean();
+
+    if ($this->open(FALSE, TRUE)) {
+      foreach ($headers as $header) {
+        // To prevent HTTP header injection, we delete new lines that are
+        // not followed by a space or a tab.
+        // See http://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2
+        $header['value'] = preg_replace('/\r?\n(?![\t ])/', '', $header['value']);
+        backdrop_add_http_header($header['key'], $header['value']);
+      }
+      // Transfer file in 1024 byte chunks to save memory usage.
+      while ($data = $this->read(1024)) {
+        print $data;
+      }
+      $this->close();
+    }
+    else {
+      backdrop_not_found();
+    }
+
+    // Start buffering and throw away the results so that errors don't get
+    // appended to the file.
+    ob_start('_backup_file_dispose_buffer');
```

Comment:
So, this _backup_file_dispose_buffer is supposed to be a callback, but it doesn't exist, so the "buffer" is thrown away? Shouldn't the callback be `NULL` in this case? Not sure.

Reply:
I'll have to look into what this does. Here's the matching code in B&M:
```
/**
 * An output buffer callback which simply throws away the buffer instead of sending it to the browser.
 */
function _backup_migrate_file_dispose_buffer($buffer) {
  return "";
}
```

Reply:
Or, maybe we don't need function transfer() at all, as we don't want to provide file download, anyway. If I got things right...

Reply:
I removed this functionality since we're not using it in the initial version.